### PR TITLE
[PyTorch] Fix CNNs Init

### LINKED ISSUE #1917: The training process of CNN implemented by Pytorch may have something wrong on the master branch.
In [7.6.2](http://preview.d2l.ai.s3-website-us-west-2.amazonaws.com/d2l-en/master/chapter_convolutional-neural-networks/lenet.html#training), [8.1.3](http://preview.d2l.ai.s3-website-us-west-2.amazonaws.com/d2l-en/master/chapter_convolutional-modern/alexnet.html#training), [8.2.3](http://preview.d2l.ai.s3-website-us-west-2.amazonaws.com/d2l-en/master/chapter_convolutional-modern/vgg.html#training), [8.3.3](http://preview.d2l.ai.s3-website-us-west-2.amazonaws.com/d2l-en/master/chapter_convolutional-modern/nin.html#training), and [8.4.3](http://preview.d2l.ai.s3-website-us-west-2.amazonaws.com/d2l-en/master/chapter_convolutional-modern/googlenet.html#training), the loss of CNN implemented by Pytorch has not decreased at all or has not decreased in the first few epochs, which is different from the results of Mxnet and Tensorflow.

### PR DESCRIPTION
**Pull request by AnirudhDagar**
*Description of changes:*
Closes #1917 

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

**Comment by mli** [comment]
Job d2l-en/PR-2000/1 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-2000/

**Comment by astonzhang** [comment]
Oh it's the 2000th PR :)

**Review by astonzhang — COMMENTED** [review]
Thanks! A few questions:

1. Why using `xavier_uniform`? What's the default init if not specified?
2. In MXNet, we specify `self.net.initialize(init.Xavier())`. Is passing the `nn.init.xavier_uniform_` handle in the arguments of `init_cnn_weights` more flexible?
3. Have we fixed all CNNs? For example, DenseNet and TextCNN both use CNNs. A global search for nn.Linear and nn.Conv2d may help.
4. Do we only need to fix CNNs? Note that `nn.Linear` is also checked. How about default inits for RNNs, MLPs, Transformers, etc.? Do we also need to apply similar fixes to them?

File: chapter_appendix-tools-for-deep-learning/utils.md [hunk 1]
```diff
@@ -261,6 +261,13 @@ def evaluate_accuracy_gpu(net, data_iter, device=None): #@save
     return metric[0] / metric[1]
 
 
+#@save
+def init_cnn_weights(m):
```
[the remark this answers is not in the document]
Comment (AnirudhDagar) [reply chapter_appendix-tools-for-deep-learning/utils.md:265]: Let's change this to `layer`. `m` was chosen earlier because of how [pytorch documentation](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.apply) shows it as a submodule.

**Comment by AnirudhDagar** [comment]
> 1. Why using xavier_uniform? What's the default init if not specified?

Default init is different for every layer in PyTorch. Eg. [`nn.Linear` init docs](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html#torch.nn.Linear) and [`nn.Conv2d` init docs](https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html#torch.nn.Conv2d) They are not uniform init, not kaiming init but something in between. Xavier Init was used in mxnet implementation as well during the CNN sections. By default `net.initialize(init.Xavier())` in MXNet will use `uniform` with `avg` factor. See [older mx1.5 detailed docs](https://mxnet.apache.org/versions/1.5.0/api/python/optimization/optimization.html#mxnet.initializer.Xavier) and [newer mx1.8 brief docs](https://mxnet.apache.org/versions/1.8.0/api/python/docs/api/initializer/index.html#mxnet.initializer.Xavier), hence we were doing the same in PyTorch where `xavier_uniform` helps achieve the same idea. See [PyTorch `xavier_uniform_` docs](https://pytorch.org/docs/master/nn.init.html#torch.nn.init.xavier_uniform_).

> 2. In MXNet, we specify self.net.initialize(init.Xavier()). Is passing the nn.init.xavier_uniform_ handle in the arguments of init_cnn_weights more flexible?

Yes but that is because MXNet offers this functionality internally whereas for PyTorch we need to implement the custom function and then pass the func to `model.apply` to apply custom weight init to the whole model. I don't think you can pass a string there, the custom function takes the submodule argument by default.

> 3. Have we fixed all CNNs? For example, DenseNet and TextCNN both use CNNs. A global search for nn.Linear and nn.Conv2d may help.

They are already using the custom init, see [Sec Defining the model in text cnn](http://preview.d2l.ai/d2l-en/master/chapter_natural-language-processing-applications/sentiment-analysis-cnn.html#defining-the-model). Also [Sec Training in Densenet](http://preview.d2l.ai/d2l-en/master/chapter_convolutional-modern/densenet.html#training) calls `train_ch6` which applied custom init already.

> 4. Do we only need to fix CNNs? Note that nn.Linear is also checked. How about default inits for RNNs, MLPs, Transformers, etc.? Do we also need to apply similar fixes to them?

We already do. Note that this addition is not new, we had custom init earlier as well, before the new API was introduced. It was just not saved like this new function that I added to the utils section (because there was no utils section previously). During the refactoring of API, this was somehow removed from the CNNs chapter. You see we already have these in different sections for stable and speedy training.

One option is not to save this as a function and define the custom function based on the section like we did earlier. In this case, I decided to save it because the whole CNNs section has the same custom function. What do you think?

**Comment by mli** [comment]
Job d2l-en/PR-2000/2 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-2000/

**Comment by astonzhang** [comment]
Thanks. I'm going to merge this PR and move the function definition of `init_cnn_weights` to the first notebook that uses it and add some descriptions there.
